compiler/rustc_lint/src/let_underscore.rs

```diff
@@ -0,0 +1,270 @@
+    fn check_local(&mut self, cx: &LateContext<'_>, local: &hir::Local<'_>) {
+        if !matches!(local.pat.kind, hir::PatKind::Wild) {
+            return;
+        }
+        if let Some(init) = local.init {
+            let init_ty = cx.typeck_results().expr_ty(init);
+            let needs_drop = init_ty.needs_drop(cx.tcx, cx.param_env);
+            let is_sync_lock = init_ty.walk().any(|inner| match inner.unpack() {
+                GenericArgKind::Type(inner_ty) => {
+                    SYNC_GUARD_PATHS.iter().any(|guard_path| match inner_ty.kind() {
+                        ty::Adt(adt, _) => {
+                            let ty_path = cx.get_def_path(adt.did());
+                            guard_path.iter().map(|x| Symbol::intern(x)).eq(ty_path.iter().copied())
+                        }
+                        _ => false,
+                    })
+                }
+
+                GenericArgKind::Lifetime(_) | GenericArgKind::Const(_) => false,
+            });
+            let is_must_use_ty = is_must_use_ty(cx, cx.typeck_results().expr_ty(init));
+            let is_must_use_func_call = is_must_use_func_call(cx, init);
+
+            if is_sync_lock {
+                cx.struct_span_lint(LET_UNDERSCORE_LOCK, local.span, |lint| {
+                    build_and_emit_lint(
+                        lint,
+                        local,
+                        init.span,
+                        "non-binding let on a synchronization lock",
+                    )
+                })
+            } else if is_must_use_ty || is_must_use_func_call {
+                cx.struct_span_lint(LET_UNDERSCORE_MUST_USE, local.span, |lint| {
+                    build_and_emit_lint(
+                        lint,
+                        local,
+                        init.span,
+                        "non-binding let on a expression marked `must_use`",
+                    );
+                })
+            } else if needs_drop {
+                cx.struct_span_lint(LET_UNDERSCORE_DROP, local.span, |lint| {
+                    build_and_emit_lint(
+                        lint,
+                        local,
+                        init.span,
+                        "non-binding let on a type that implements `Drop`",
+                    );
+                })
+            }
+        }
+
+        fn build_and_emit_lint(
+            lint: LintDiagnosticBuilder<'_, ()>,
+            local: &hir::Local<'_>,
+            init_span: rustc_span::Span,
+            msg: &str,
+        ) {
+            lint.build(msg)
+                .span_suggestion_verbose(
+                    local.pat.span,
+                    "consider binding to an unused variable",
+                    "_unused",
+                    Applicability::MachineApplicable,
+                )
+                .span_suggestion_verbose(
+                    init_span,
+                    "consider explicitly droping with `std::mem::drop`",
+                    "drop(...)",
```

Comment:
Should we craft an applicable suggestion here, or do we expect low usefulness?

Reply:
I don't know how to do this--I tried using `format!("drop({})", init)`  to create an applicable suggestion, but I couldn't figure out how to turn the `init` expression into text.

Reply:
Update: I was able to figure out how to create a suggestion using `multipart_suggestion`. I wasn't sure how to remove the `let _ =` part of the message however (In the docs, I can't seem to figure out where the `let` and `=` parts of a `Local` statement are kept in the span)